src/replication/replication_manager.cpp

```diff
@@ -0,0 +1,290 @@
+  bool completed = false;
+  try {
+    messenger_->SendMessage(
+        GetReplicaConnection(replica_name), msg,
+        [this, block, &completed](common::ManagedPointer<messenger::Messenger> messenger,
+                                  const messenger::ZmqMessage &msg) {
+          if (block) {
+            // If this isn't a blocking send, then completed will fall out of scope.
+            completed = true;
+            blocking_send_cvar_.notify_all();
+          }
+        },
+        static_cast<uint64_t>(type));
+
+    if (block) {
+      std::unique_lock<std::mutex> lock(blocking_send_mutex_);
+      // If the caller requested to block until the operation was completed, the thread waits.
+      if (!blocking_send_cvar_.wait_for(lock, REPLICATION_MAX_BLOCKING_WAIT_TIME, [&completed] { return completed; })) {
+        // TODO(WAN): Additionally, this is hackily a messenger exception so that it gets handled by the catch.
+        throw MESSENGER_EXCEPTION("TODO(WAN): Handle a replica dying in synchronous replication.");
+      }
+      lock.unlock();
+    }
+  } catch (const MessengerException &e) {
+    REPLICATION_LOG_WARN(fmt::format("[FAILED] Send -> {} (type {} block {}): msg size {}", replica_name,
+                                     static_cast<uint8_t>(type), block, msg.size()));
+  }
+}
+
+void ReplicationManager::ReplicateBuffer(storage::BufferedLogWriter *buffer) {
+  if (!replication_enabled_) {
+    REPLICATION_LOG_WARN(fmt::format("Skipping replicate buffer as replication is disabled."));
+    return;
+  }
+
+  NOISEPAGE_ASSERT(buffer != nullptr,
+                   "Don't try to replicate null buffers. That's pointless."
+                   "You might plausibly want to track statistics at some point, but that should not happen here.");
+  // TODO(WAN): Is it true that a replica will never want to send its buffers to the primary?
+  // TODO(WAN): Cache the identity check into a bool?
+  if (IsPrimary()) {
+    common::json j;
+    // TODO(WAN): Add a size and checksum to message.
+    j["buf_id"] = next_buffer_sent_id_++;
+    j["content"] = nlohmann::json::to_cbor(std::string(buffer->buffer_, buffer->buffer_size_));
+
+    for (const auto &replica : replicas_) {
+      // TODO(WAN): many things break when block is flipped from true to false.
+      ReplicaSend(replica.first, MessageType::REPLICATE_BUFFER, j.dump(), true);
+    }
+  }
+
+  if (buffer->MarkSerialized()) {
+    empty_buffer_queue_->Enqueue(buffer);
+  }
+}
+
+void ReplicationManager::EventLoop(common::ManagedPointer<messenger::Messenger> messenger,
+                                   const messenger::ZmqMessage &msg) {
+  switch (static_cast<MessageType>(msg.GetDestinationCallbackId())) {
+    case MessageType::ACK: {
+      nlohmann::json json = nlohmann::json::parse(msg.GetMessage());
+      REPLICATION_LOG_TRACE(fmt::format("ACK: {}", json.dump()));
+      auto callback_id = json.at("callback_id").get<uint64_t>();
```

Comment:
This applies to all the `json.at(...)` calls in this PR, but I think it would be better to extract the JSON keys out to some constants somewhere.

Reply:
Refactored to do the struct which contains constants.